src/lib/libled.c

```diff
@@ -98,6 +99,50 @@ led_status_t led_scan(struct led_ctx *ctx)
 	return ctx->deferred_error;
 }
 
+static bool virt_nvme(const char *path)
+{
+	if (strstr(path, "/sys/devices/virtual/nvme-subsystem"))
+		return true;
+	return false;
+}
+
+static led_status_t get_nvme_controller(char *virtual_blockdev, char *result)
+{
+	/*
+	 * Converts this
+	 * /sys/devices/virtual/nvme-subsystem/nvme-subsys0/nvme0n1
+	 * to
+	 * /sys/devices/pci0000:64/0000:64:02.0/0000:65:00.0/0000:66:00.0/0000:67:00.0/
+	 * 0000:68:04.0/0000:6a:00.0/nvme/nvme0/nvme0c0n1
+	 *
+	 * In NVMeOF environments, this process will result in a virtual path
+	 */
+
+	led_status_t rc = LED_STATUS_INVALID_PATH;
+	glob_t results;
+	int nvme_num, ns;
+	char pattern[PATH_MAX];
+
+	int ret = sscanf(virtual_blockdev,
+		"/sys/devices/virtual/nvme-subsystem/nvme-subsys%*d/nvme%dn%d",
+		&nvme_num, &ns);
+
+	if (ret != 2)
+		return rc;
+
+	ret = snprintf(pattern, PATH_MAX, "/sys/block/nvme%dn%d/device/nvme*/nvme*c*n%d",
+			nvme_num, ns, ns);
+	if (ret >= PATH_MAX)
+		return rc;
+
+	ret = glob(pattern, 0, NULL, &results);
```

Comment:
The device should be on block list, can we go through block_devices instead?

Reply:
> The device should be on block list, can we go through block_devices instead?

This change is to translate from `/dev/nvme1n1` which is virtual to something which is physical in sysfs as only physical devices are the ones in the block list we have in memory.

Reply:
It seems that for example: /dev/nvme7c7n1 is on block list. So instead of searching sysfs we can got the device from block list:
```
lrwxrwxrwx 1 root root 0 Aug 23 17:24 nvme7c7n1 -> ../devices/pci0000:ae/0000:ae:05.5/pci10001:00/10001:00:02.0/10001:08:00.0/10001:09:00.0/10001:0a:00.0/nvme/nvme7/nvme7c7n1
lrwxrwxrwx 1 root root 0 Aug 23 17:24 nvme7n1 -> ../devices/virtual/nvme-subsystem/nvme-subsys7/nvme7n1

```
I will double verify it on morning.

Reply:
> It seems that for example: /dev/nvme7c7n1 is on block list. So instead of searching sysfs we can got the device from block list:

There is no `/dev/nvme7c7n1`  ,  there is a `/dev/nvme7n1` which is what the end users can see and interact with.  To translate from `/dev/nvme7n1` to the sysfs path that refers to `nvme7c7n1` which is in the block list is the root problem.  I looked around in sysfs to see if I could find a better way to do this, but ultimately what this PR does is outlined in: https://lore.kernel.org/all/3c725e5deaabaaf145f48f2f6fcfdae9f6d41e2e.camel@suse.de/t/

If you can come up with a better way to do this, please do as I really don't like this approach.

Intel makes dual ported U.2 NVMe devices, maybe you could get access to some of them in your lab?  It would be good to have hardware that has multiple physical paths to the NVMe storage device for testing to ensure the changes made work in that configuration.

Reply:
> There is no `/dev/nvme7c7n1` , there is a `/dev/nvme7n1` which is what the end users can see and interact with. To translate from `/dev/nvme7n1` to the sysfs path that refers to `nvme7c7n1` which is in the block list is the root problem. I looked around in sysfs to see if I could find a better way to do this, but ultimately what this PR does is outlined in: https://lore.kernel.org/all/3c725e5deaabaaf145f48f2f6fcfdae9f6d41e2e.camel@suse.de/t/

Now I understand, we cannot simply assume that nvme7n1 us nvme7c7n1 because real dual-port drive there could have 2 links, like in your example:
> ls -d /sys/block/nvme0n${NSID}/device/nvme*/nvme*c*n${NSID}
> /sys/block/nvme0n1/device/nvme0/nvme0c0n1 
> /sys/block/nvme0n1/device/nvme1/nvme0c1n1 

But... the simple matching will work in 90% cases. I don't see a strong need to support real dual-pathing. My product, does not support dual-port connection, so for me it is acceptable to:
-  assume that "nvme7n1" is "nvme7c7n1"
- return real path to this device from block list or return realpath(/sys/block/nvme7c7n1)

I understand that it is quite incorrect with spec but I'm fine with that- as I said I don't need to support dual-port connections. I need to just understand multipathing in base case.
I think that your implementation doesn't cover it well too, because you are always using `glob[0]` result:
- let imagine dual port nvme0n1 devnode with two controllers nvme0 (nvme0c0n1) and nvme1 (nvme0c1n1)
- let say that nvme0c1n1 is the one supported by ledmon
- we are looking for the first glob result so the led management is not supported anyway (if it is nvme0c0n1)
- after the reboot the nvme devices may appear in different order (unlikely but they can), so after the reboot it may start to work.

We could have different results in extraordinary cases (like changing kernel).  I don't know and enclosure with dual port support (one port will be always not supported?) The globing will help only when one controller will disappear during runtime, so the application will automatically switch to second controller (if supported). However these could be two physical locations so we will start to blink other led. I would say that better is to use the one controller every time.

I think that in this case we should set led state to all supported controllers but it requires full implementation in lib/ledmon/ledctl, not just translation layer.

> If you can come up with a better way to do this, please do as I really don't like this approach.
Let me know what do you think about simplification I described above. I think that it is all we need now.

The full support of MP is something I need to deeply reconsider, but here draft I can imagine now:
I need to handle it from the beginning, so the device represented by multipath should be added to block list. "nvmeXcXnY" should not appear on block list. At this time it will be reasonable to change `led_device_name_lookup` to return path to orginal devnode, not a full path, we should do matching by devnodes. Then, I will be ready to keep many paths to one block_device (dual-port connection) and when set_led is requested It will go through all paths and set_state.

Just an idea, nothing decided yet. It is something that needs to be projected but it needs to wait (currently, I'm working on NPEM kernel driver).

Hope it has some sense to you!
Thanks,
Mariusz

Reply:
> return real path to this device from block list or return realpath(/sys/block/nvme7c7n1)

Wow, it shocked me that /dev/nvme6c6n1 is not visible in /dev directory, it was not something I expected. I thought that all devices in /sys/block have nodes in /dev/.

```
[root@gklab-108-180 ledmon_gh_master]# ll /dev/ | grep nvme6
crw-------   1 root root    243,   6 Aug 23 17:24 nvme6
brw-rw----   1 root disk    259,  16 Aug 23 17:24 nvme6n1

```

This works for me, it think some soft of it is enough we need for now.

```diff
diff --git a/src/lib/libled.c b/src/lib/libled.c
index 7d32cf4..1979d12 100644
--- a/src/lib/libled.c
+++ b/src/lib/libled.c
@@ -119,7 +119,6 @@ static led_status_t get_nvme_controller(char *virtual_blockdev, char *result)
         */

        led_status_t rc = LED_STATUS_INVALID_PATH;
-       glob_t results;
        int nvme_num, ns;
        char pattern[PATH_MAX];

@@ -130,26 +129,25 @@ static led_status_t get_nvme_controller(char *virtual_blockdev, char *result)
        if (ret != 2)
                return rc;

-       ret = snprintf(pattern, PATH_MAX, "/sys/block/nvme%dn%d/device/nvme*/nvme*c*n%d",
-                       nvme_num, ns, ns);
+       ret = snprintf(pattern, PATH_MAX, "/sys/block/nvme%dc%dn%d",
+                       nvme_num, nvme_num, ns);
        if (ret >= PATH_MAX)
                return rc;

-       ret = glob(pattern, 0, NULL, &results);
-       if ((ret == 0) && (results.gl_pathc > 0) && (realpath(results.gl_pathv[0], result)))
-               rc = LED_STATUS_SUCCESS;
+       realpath(pattern, result);

-       globfree(&results);
-       return rc;
+       return 0;
 }


```

For the reasons I raised above I think that we don't need to care about dual port connections and we can assume that the device will be named this way. I can analyze nvme stack in kernel if you see that as a dangerous. I think that it is fine.